### PR TITLE
FSE: filter out templates that have missing Blocks

### PR DESCRIPTION
**Pull request by getdave**
#### Changes proposed in this Pull Request

Short term solution for https://github.com/Automattic/wp-calypso/issues/39293 whilst we work on [improved longer term solutions](https://github.com/Automattic/wp-calypso/issues/39293#issuecomment-595147085).

[Recent Issues](https://github.com/Automattic/wp-calypso/issues/39293) have shown that users are seeing missing block messages in Templates when these Templates are using blocks that are not registered.

The missing blocks are likely due to certain blocks being used which are not directly a part of the FSE plugin (eg: [`layout-grid`](https://github.com/Automattic/wp-calypso/issues/39293)).

This PR proposes a solution whereby we dynamically remove any Templates from SPT which contain unsupported Blocks. 

## Todo

- [x] Flatten Blocks before checking for missing in order to catch nested Blocks
- [x] Update testing `MISSING` block with actually missing block name `core/missing`(?)
- [x] Work out why thumbnails are still showing even when filtered out (try clicking them - you'll see they don't load cos template is not available!)

## Testing instructions

You will need to be able to build and then sync the FSE Plugin over to a WP install. Full instructions can be found at `PaYJgx-sW-p2`.

### Testing Setup

* Disable Jetpack.
* Create a new Page 
* Open developer tools and open the `Console` tab.
* Type the following and hit ENTER:
```
Boolean(wp.blocks.getBlockTypes().find(block => block.name === "jetpack/layout-grid"));
``` 
* You should see `false` if the experimental block `jetpack/layout-grid` is not available.

**Warning**: if you see `true` then disable the "Block Experiments" Plugin from your env and rerun the command.

### On Master

* Switch to `master` branch.
* Build FSE and sync to your WP install.
* Create a new Page and see the SPT Template Selector appear.
* You should see the About template showing warning notices about the missing `jetpack/layout-grid` Plugin (see screenshot below): 
Screen Shot 2020-03-04 at 16 48 13

### On this PR's branch

* Checkout this PR's branch.
* Build FSE and sync to your WP install.
* Create a new Page and see the SPT Template Selector appear.
* You should no longer see the "About" template which uses the `jetpack/layout-grid` Block. 
Screen Shot 2020-03-04 at 16 50 21

* Check all the other templates that use the `jetpack/layout-grid` Block do not show up in the Template selector (you can [use @marekhrabe's analysus to see which Templates use the Block](https://gist.github.com/marekhrabe/1ca4d5f436040afbced1123444088d7d#file-stats-json-L55))
* Click through all the templates that do show and make sure none are showing the "missing blocks" warnings. 

Fixes https://github.com/Automattic/wp-calypso/issues/39293

### CODE REVIEW

**Comment by mkaz** [comment]
I like this approach, it is capability testing vs. version number testing. 

Would it be possible to include messaging around templates that are filtered out? For example, "You are not saying all available templates because you have deactivated XYZ plugin. Enable the plugin (link) to see more templates"

**Comment by getdave** [comment]
> I like this approach, it is capability testing vs. version number testing.

Thanks!

> Would it be possible to include messaging around templates that are filtered out? For example, "You are not saying all available templates because you have deactivated XYZ plugin. Enable the plugin (link) to see more templates"

I think that takes us into the realm of detecting specific Plugins. Personally, I think this PR should stay focused on fixing the broken user experience.

Perhaps a follow-up PR could add more features such as you suggest. For example, I was thinking we could even show "ghost" versions of the Templates which are not clickable and then when you hover it prompts you to install a Plugin to access those Templates.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js [hunk 1]
```diff
@@ -289,13 +315,28 @@ class PageTemplateModal extends Component {
 			return null;
 		}
 
+		// The raw `templates` prop is not filtered to remove Templates that
```
Comment (getdave) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: @marekhrabe I wonder can you think of a better way to filter the `this.props.templates`? Remember this is the raw blocks before they are parsed and therefore it doesn't have `core/missing`. This is why I'm relying on the `getBlocksByTemplateSlugs` which has been filtered to then filter the `this.props.templates`. Feels a tad brittle though.

**Comment by marekhrabe** [comment]
Screenshot 2020-03-05 at 11 06 35

Templates with unsupported content are gone with this PR ✅

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js [hunk 2]
```diff
@@ -40,8 +41,8 @@ class PageTemplateModal extends Component {
 	);
 
 	// Parse templates blocks and memoize them.
-	getBlocksByTemplateSlugs = memoize( templates =>
-		reduce(
+	getBlocksByTemplateSlugs = memoize( templates => {
+		let rtn = reduce(
 			templates,
 			( prev, { slug, content, title } ) => {
```
Comment (marekhrabe) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: Can we name these so they are more self-explanatory? Both `rtn` and `prev` (I know the latter one was not introduced in this PR) feel like they are shortened for no reason and explaining the function call details or how reduce internals work but not really the data they hold — eg: templates, templatesBySlug or something

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js [hunk 3]
```diff
@@ -62,8 +63,33 @@ class PageTemplateModal extends Component {
...
 			},
 			{}
-		)
-	);
+		);
+
+		// Remove templates that include a missing block
+		rtn = this.filterTemplatesWithMissingBlocks( rtn );
+
+		return rtn;
```
Comment (marekhrabe) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]:
```suggestion
		return this.filterTemplatesWithMissingBlocks( rtn );
```

I think we can simplify this and switch the variable to const.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+const MISSING_BLOCK_NAME = 'core/missing';
+
+/**
+ * Determines whether the provided collection of Blocks contains any "missing"
+ * blocks as determined by the presence of the `core/missing` block type.
+ *
+ * @param {Array} blocks the collection of block objects to check for "missing" block .
+ * @returns {boolean} whether the collection blocks contains any missing blocks.
+ */
+function containsMissingBlock( blocks ) {
```
Comment (marekhrabe) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: Nice and clean solution to extract the function like into its file! 👍

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js (hunk 1 above)
Comment (marekhrabe) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: I've been thinking about this and for the sake of shipping this fix, I would not further complicate this PR with some sort of refactor. I have some ideas in my mind and I will document them in an issue that we might or might not pick up in future.

tldr: let's keep this approach for now, nothing really wrong with it

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js (hunk 2 above)
Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: Agreed. Fixed.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js (hunk 3 above)
Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: I fixed this as part of the other comment.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js (hunk 1 above)
Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js]: I agreed. Also, have ideas. Please do create that follow up refactor Issue if you can.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js (hunk 4 above)
Comment (retrofox) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: minor: does it make sense returning immediately if the blocks array is empty?

```es6
if ( ! blocks || ! blocks.length ) {
  return [];
}
```

Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: I see why you are asking. However, the purpose of the function is to detect whether the blocks "contains the missing block". An empty set of blocks _does not_ contain the missing block. Therefore IMHO it is valid.

Do you have a particular concern about it I might have missed?

Also, note that the blank template doesn't have any blocks so if we didn't account for this then the blank template would never show. However, note that this early return is not directly linked to the blank template - it's just a nice side effect.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	return blocks.find( block => {
```
Comment (retrofox) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: minor: could you cast the returning value, adding `!!` of something like that.
It supposes to be a boolean, however, it's either a block object or `undefined`
Or update the jsdoc. 

```es6
return !! blocks.find( block => {
  // ...
}
```

Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: I agree. It's better to be explicit about the type here.

Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: Fixed.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js (hunk 4 above)
Comment (retrofox) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: I just was testing this code and saw that the blocks array was empty. So in order to avoid calling the `find()` function, return `false` immediately. That was my point. However, I prelabeled with `minor`, so feel free to ignore it :-D

**Comment by getdave** [comment]
@vindl confirmed that it is ok to merge build of FSE as this is now legacy and won't be included on new PRs by default. 

Waiting on @davemart-in to confirm he's happy with this PR to merge given discussions over at https://github.com/Automattic/wp-calypso/issues/39293

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js (hunk 4 above)
Comment (getdave) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: Oh I see! This code isn't there and this is a suggestion. Whoops! Yes we should add it. My bad!

**Comment by vindl** [comment]
> @vindl confirmed that it is ok to merge build of FSE as this is now legacy and won't be included on new PRs by default.

Are you referring to failed FSE build in checks? It's not legacy but a new addition, but its failure shouldn't stop you from merging, since this check is not being run on master branch.

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * Determines whether the provided collection of Blocks contains any "missing"
+ * blocks as determined by the presence of the `core/missing` block type.
+ *
+ * @param {Array} blocks the collection of block objects to check for "missing" block .
+ * @returns {boolean} whether the collection blocks contains any missing blocks.
+ */
+function containsMissingBlock( blocks ) {
+	return !! blocks.find( block => {
+		// If we found a missing block the bale out immediately
```
Comment (marekhrabe) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/contains-missing-block.js]: I've never seen the word "bale" but seems it's valid :) https://brians.wsu.edu/2016/05/19/bail-bale/

**Comment by davemart-in** [comment]
A couple of things:

One last minor thing: In this image, can we eliminate the empty "Blog pages" section header and any other empty section headers?

75971062-845bb580-5ed1-11ea-8a6f-4beff964b0b4

For a follow on v2, I the combination of what @mkaz and @getdave came up with:

> **Mkaz** - Would it be possible to include messaging around templates that are filtered out? For example, "You are not saying all available templates because you have deactivated XYZ plugin. Enable the plugin (link) to see more templates"

> **getdave** - Perhaps a follow-up PR could add more features such as you suggest. For example, I was thinking we could even show "ghost" versions of the Templates which are not clickable and then when you hover it prompts you to install a Plugin to access those Templates.

I'll pull together a design for this and we can get started on it immediately after this PR ships.

**Comment by getdave** [comment]
> One last minor thing: In this image, can we eliminate the empty "Blog pages" section header and any other empty section headers?

@davemart-in Not at all minor. Great catch. A flaw in my logic!  Fixed.

Screen Shot 2020-03-06 at 15 30 56

> For a follow on v2, I the combination of what @mkaz and @getdave came up with:

This would be good. I have a fair idea of how this could work already. Looking forward to your designs.
